pkg/storegateway/bucket.go

```diff
@@ -849,6 +853,17 @@ func (s *BucketStore) sendStreamingChunks(
 			batchSizeBytes = 0
 		}
 
+		if !haveSentEstimatedChunks && chunksCount > 0 {
```

Comment:
I think the implementation on the querier side is very complicated because the estimations may arrive after 1+ batches already sent. I think we should send the estimations **before** we send any batch. Why not moving this block above `for _, sc := range set.series {`? We can iterate `set.series` once to compute the number of chunks and then send the estimation. If the batch has no chunks, we send an estimation of 0, so that we guarantee the estimation is always sent before the first batch.

Reply:
Good point, I'm not sure why I did it this way (I suspect I'm too used to forward-only iterators everywhere 🤷).